app/src/main/java/org/p2p/wallet/auth/repository/CountryCodeInMemoryRepository.kt

```diff
@@ -1,30 +1,129 @@
 package org.p2p.wallet.auth.repository
 
-import org.p2p.wallet.auth.ui.phone.CountryPickerParser
+import android.content.Context
+import android.telephony.TelephonyManager
+import io.michaelrocks.libphonenumber.android.PhoneNumberUtil
+import org.p2p.wallet.R
 import org.p2p.wallet.auth.model.CountryCode
+import org.p2p.wallet.common.di.AppScope
+import org.p2p.wallet.infrastructure.dispatchers.CoroutineDispatchers
+import org.p2p.wallet.utils.emptyString
+import org.xmlpull.v1.XmlPullParser
+import org.xmlpull.v1.XmlPullParserFactory
+import timber.log.Timber
 
 class CountryCodeInMemoryRepository(
-    private val parser: CountryPickerParser
+    private val phoneNumberUtil: PhoneNumberUtil,
+    private val appScope: AppScope,
+    private val dispatchers: CoroutineDispatchers,
+    private val context: Context
 ) : CountryCodeLocalRepository {
 
-    private var countryCodes = mutableListOf<CountryCode>()
+    private val countryCodes = mutableListOf<CountryCode>()
     private var countryCodeMask = mutableMapOf<String, String>()
 
-    override suspend fun getCountryCodes(): List<CountryCode> {
+    private val KEY_COUNTRY = "country"
+    private val KEY_NAME_CODE = "name_code"
+    private val KEY_PHONE_CODE = "phone_code"
+    private val KEY_NAME = "name"
+    private val KEY_FLAG_EMOJI = "flag_emoji"
+
+    override suspend fun getCountryCodes(): List<CountryCode> = countryCodes
+
+    override suspend fun detectCountryCodeByLocale(): CountryCode? {
+        return try {
+            val localeCountryIso = context.resources.configuration.locale.country
+            getCountryForISO(localeCountryIso)
+        } catch (e: Exception) {
+            null
+        }
+    }
+
+    override suspend fun detectCountryCodeByNetwork(): CountryCode? {
+        return try {
+            val telephonyManager = context.getSystemService(TelephonyManager::class.java)
+            val networkCountryIso = telephonyManager.networkCountryIso
+            getCountryForISO(networkCountryIso)
+        } catch (e: Exception) {
+            null
+        }
+    }
+
+    override suspend fun detectCountryCodeBySimCard(): CountryCode? {
+        return try {
```

Comment:
repository is responsible for making work on IO thread, place withContext here please